### ISSUE TITLE
Cutadapt sample name parsing too simple. Option to use filename?

### ISSUE THREAD

**Issue by biodray** [ewels/MultiQC#949]
**Description of bug:**
I'm running MultiQC in a folder containing many cutadapt log files. However, only one file is used (the last one), even if I specified one by one every files I want it to use.

**MultiQC Error log:**
There's not error log entry

**File that triggers the error:**
[TE18-018_12S_cut_log.txt](https://github.com/ewels/MultiQC/files/3069409/TE18-018_12S_cut_log.txt)
[TE18-019_12S_cut_log.txt](https://github.com/ewels/MultiQC/files/3069410/TE18-019_12S_cut_log.txt)

**MultiQC run details:**
 - Command used to run MultiQC: `multiqc [list of files names] --module cutadapt -f`
 - MultiQC Version: MultiQC v 1.7
 - Operating System: Ubuntu 18.04.2 LTS
 - Python Version: Python 2.7.15rc1
 - Method of MultiQC installation: pip

**Comment by ewels** [comment]
Hi @biodray,

Have you had a read of the docs on this subject? See https://multiqc.info/docs/#not-enough-samples-found

Let me know if these don't fix the problem for you and we can reopen the issue.

Phil

**Comment by biodray** [comment]
Hi @ewels ,

Thanks for your anwser, sadly it didn't fix my problem. A closer look within the html report make me realize that multiQC think all my files have one sample with the exact name "--discard-untrimmed" (which is an argument of cutadapt), so it is overwriting the results instead of aggregating them. So at the end only the last file is used ...

**Comment by ewels** [comment]
Hi @biodray,

The docs that I linked to cover exactly this scenario (specifically https://multiqc.info/docs/#clashing-sample-names ), there are several ways to modify the MultiQC behaviour to avoid identical sample names. Probably best here is to use sensibly named subdirectories. Not ideal I agree, but it should work.. 😄 

However - using a cutadapt flag instead of a filename sounds like buggy behaviour. Looking at the code, it seems MultiQC just takes the final part of the cutadapt command: https://github.com/ewels/MultiQC/blob/2037d6322b2554146a74efbf869156ad20d4c4ec/multiqc/modules/cutadapt/cutadapt.py#L104-L105

I'm not sure of how else we can do this automatically, though this is another case similar to issues #864 and #890 where it could be nice to have an option to disable the automatic sample name behaviour and just use the input filenames.

**Comment by ronin-gw** [comment]
Just sharing my case: I know current module handling [the case where input is read from stdin](https://github.com/ewels/MultiQC/blob/master/multiqc/modules/cutadapt/cutadapt.py#L106-L108), but I got a similar problem when I used process substitution as input.

In this cutadapt log example, MultiQC cathes `13` as a sample name.
```
This is cutadapt 1.15 with Python 3.5.2
Command line parameters: -a A{100} -m 20 -q 20,20 -b file:/proc/self/fd/12 /proc/self/fd/13
```

**Comment by daler** [comment]
For anyone else running across this, a temporary workaround is to ensure that the original cutadapt call *ends* with the input filename:

```bash
# wrong, sample will show up as "20"
cutadapt sample.fastq.gz -a AAAAA -q 20 

# correct, sample will show up as "sample.fastq.gz"
cutadapt -a AAAAA -q 20 sample.fastq.gz
```

**Comment by ewels** [comment]
x-ref to #1482 where a new cutadapt JSON summary output is discussed. This should resolve the problems described here about the order of command line arguments, as the input paths are explicitly described with their own JSON keys (no need to guess from the command used).

I have been thinking a bit about whether I can still fix this for the previous style of logs, but it's deceptively tricky. To do it properly, I think I'd have to know all possible cutadapt cli option flags and whether they take an argument or not. Then I could clean them away to leave just the positional arguments (the input files). This is tricky and would need a fair bit of maintenance, which I'm not keen on.

Another approach that I could do is to work back from the end and ignore anything that looks like a cli flag. That will solve the `--discard-untrimmed` example, but it won't solve the `-q 20` example. But it's probably better than nothing.

In summary then, I think I will take 3 actions here:

- [ ] Chew back from the end until I find a string that doesn't start with `-` or `--`
- [ ] Use the new JSON output to do this in a better way in the future
- [ ] Add a new central MultiQC feature to use the filename as the sample name, overriding any module that tries to be more clever

**Comment by ewels** [comment]
First one done - @biodray your example files now work for me as you'd expect. I now get `TE18-018_12S_R2` and `TE18-019_12S_R2` as sample names instead of `-discard-untrimmed`... 👍🏻

**Comment by ewels** [comment]
Ok, and a new command line option `--fn_as_s_name` for you @ronin-gw (sorry for the delay).

Documentation here: https://multiqc.info/docs/#using-log-filenames-as-sample-names

Pasted here for convenience:

> ## Using log filenames as sample names
> 
> A substantial number of MultiQC modules take the sample name identifiers that you see in the report from the file contents - typically the filename of the input file. This is because log files can often be called things like `mytool.log` or even concatenated. Using the input filename used by the tool is typically safer and more consistent across modules.
> 
> However, sometimes this does not work well. For example, if the input filename is not relevant (eg. using a temporary file or FIFO, process substitution or stdin etc.). In these cases your log files may have useful filenames but MultiQC will not be using them.
> 
> To force MultiQC to use the log filename as the sample identifier, you can use the `--fn_as_s_name` command line flag or set the `use_filename_as_sample_name`:
> 
> ```yaml
> use_filename_as_sample_name: true
> ```
> 
> This affects all modules and all search patterns. If you want to limit this to just one or more specific search patterns, you can do by giving a list:
> 
> ```yaml
> use_filename_as_sample_name:
>   - cutadapt
>   - picard/gcbias
>   - picard/markdups
> ```
> 
> Note that this should be the search pattern key (see above) and not just the module name. This is because some modules search for multiple files.
> 
> The log filename will still be cleaned. To use the raw log filenames, combine with the `--fullnames`/`-s` flag or `fn_clean_sample_names` config option described above.

I hope this is helpful! Sorry for the delay. Once @marcelm is decided on a JSON structure for the new summary files, I can add support for those and tidy up the final of the 3rd check boxes for a proper solution 👍🏻
